commands/new_function_test.go

```diff
@@ -5,37 +5,180 @@ package commands
+	var funcYAML string
+	funcYAML = funcName + ".yml"
 
 	// Cleanup the created directory
 	defer func() {
 		os.RemoveAll(funcName)
-		os.Remove(funcName + ".yml")
-		os.Remove("template")
+		os.Remove(funcYAML)
 	}()
 
+	cmdParameters := []string{
+		"new",
+		funcName,
+		"--lang=" + funcLang,
+		"--gateway=" + defaultGateway,
+	}
+
 	stdOut := test.CaptureStdout(func() {
-		faasCmd.SetArgs([]string{
-			"new",
-			funcName,
-			"--lang=python",
+		faasCmd.SetArgs(cmdParameters)
+		faasCmd.Execute()
+	})
+
+	// Validate new function output
+	if found, err := regexp.MatchString(nft.expectedMsg, stdOut); err != nil || !found {
+		t.Fatalf("Output is not as expected: %s\n", stdOut)
+	}
+
+	if nft.expectedMsg == SuccessMsg {
+
+		// Make sure that the folder and file was created:
+		if _, err := os.Stat("./" + funcName); os.IsNotExist(err) {
+			t.Fatalf("%s/ directory was not created", funcName)
+		}
+
+		if _, err := os.Stat(funcYAML); os.IsNotExist(err) {
+			t.Fatalf("\"%s\" yaml file was not created", funcYAML)
+		}
+
+		// Make sure that the information in the YAML file is correct:
+		parsedServices, err := stack.ParseYAMLFileForStack(funcYAML, "", "")
+		if err != nil {
+			t.Fatalf("Couldn't open modified YAML file \"%s\" due to error: %v", funcYAML, err)
+		}
+		services := *parsedServices
+
+		var testServices stack.Services
+		testServices.Provider = stack.Provider{Name: "faas", GatewayURL: defaultGateway}
+		if !reflect.DeepEqual(services.Provider, testServices.Provider) {
+			t.Fatalf("YAML `provider` section was not created correctly for file %s: got %v", funcYAML, services.Provider)
+		}
+
+		testServices.Functions = make(map[string]stack.Function)
+		testServices.Functions[funcName] = stack.Function{Language: funcLang, Image: funcName, Handler: "./" + funcName}
+		if !reflect.DeepEqual(services.Functions[funcName], testServices.Functions[funcName]) {
+			t.Fatalf("YAML `functions` section was not created correctly for file %s, got %v", funcYAML, services.Functions[funcName])
+		}
+	}
+
+}
+
+func Test_newFunctionTests(t *testing.T) {
+	// Change directory to testdata
+	if err := os.Chdir("testdata/new_function"); err != nil {
+		t.Fatalf("Error on cd to testdata dir: %v", err)
+	}
+
+	for _, test := range NewFunctionTests {
+		t.Run(test.title, func(t *testing.T) {
+			runNewFunctionTest(t, test)
 		})
+	}
+
+	if err := os.Chdir("../.."); err != nil {
+		t.Fatalf("Error on cd back to commands/ directory: %v", err)
+	}
+}
+
+func Test_newFunctionListCmds(t *testing.T) {
+
+	if err := os.Chdir("testdata/new_function"); err != nil {
+		t.Fatalf("Error on cd to testdata dir: %v", err)
+	}
+
+	cmdParameters := []string{
+		"new",
+		"--list",
+	}
+
+	stdOut := test.CaptureStdout(func() {
+		faasCmd.SetArgs(cmdParameters)
 		faasCmd.Execute()
+	})
+
+	// Validate new function output
+	if !strings.HasPrefix(stdOut, ListOptionOutput) {
+		t.Fatalf("Output is not as expected: %s\n", stdOut)
+	}
 
+	if err := os.Chdir("../.."); err != nil {
```

Comment:
I don't think I like this because it's cryptic. Please can you explain what's happening?

Reply:
We need to execute these tests in `faas-cli/commands/testdir/new_function` directory, so we need to `cd` into it for the test to run, then cd back to the home directory afterwards (because we don't know what order the tests will execute in).

I've changed this to save the full directory path at the beginning of the function, then `cd` back into it instead of using the relative `../..`.